synapse/lib/auth.py

```diff
@@ -0,0 +1,1301 @@
+                    m = f'Matched role rule ({rt}) for role {self.rolename} on gate {self.gateiden}.'
+                else:
+                    m = f'Matched user rule ({rt}) on gate {self.gateiden}.'
+            else:
+                if self.roleiden:
+                    m = f'Matched role rule ({rt}) for role {self.rolename}.'
+                else:
+                    m = f'Matched user rule ({rt}).'
+            return m
+
+        return 'No matching rule found.'
+
+class Auth(s_nexus.Pusher):
+    '''
+    Auth is a user authentication and authorization stored in a Slab.  Users
+    correspond to separate logins with different passwords and potentially
+    different privileges.
+
+    Users are assigned "rules".  These rules are evaluated in order until a rule
+    matches.  Each rule is a tuple of boolean, and a rule path (a sequence of
+    strings).  Rules that are prefixes of a privilege match, i.e.  a rule
+    ('foo',) will match ('foo', 'bar').
+
+    Roles are just collections of rules.  When a user is "granted" a role those
+    rules are assigned to that user.  Unlike in an RBAC system, users don't
+    explicitly assume a role; they are merely a convenience mechanism to easily
+    assign the same rules to multiple users.
+
+    AuthGates are objects that manage their own authorization.  Each
+    AuthGate has roles and users subkeys which contain rules specific to that
+    user or role for that AuthGate.  The roles and users of an AuthGate,
+    called GateRole and GateUser respectively, contain the iden of a role or
+    user defined prior and rules specific to that role or user; they do not
+    duplicate the metadata of the role or user.
+
+    Layout::
+
+        Auth root (passed into constructor)
+        ├ roles
+        │   ├ <role iden 1>
+        │   ├ ...
+        │   └ last role
+        ├ users
+        │   ├ <user iden 1>
+        │   ├ ...
+        │   └ last user
+        └ authgates
+            ├ <iden 1>
+            │   ├ roles
+            │   │   ├ <role iden 1>
+            │   │   ├ ...
+            │   │   └ last role
+            │   └ users
+            │       ├ <user iden 1>
+            │       ├ ...
+            │       └ last user
+            ├ <iden 2>
+            │   ├ ...
+            └ ... last authgate
+
+    '''
+
+    async def __anit__(self, slab, dbname, pref='', nexsroot=None, seed=None, maxusers=0):
+        '''
+        Args:
+            slab (s_lmdb.Slab): The slab to use for persistent storage for auth
+            dbname (str): The name of the db to use in the slab
+        '''
+        # Derive an iden from the db name
+        iden = f'auth:{dbname}'
+        await s_nexus.Pusher.__anit__(self, iden, nexsroot=nexsroot)
+
+        self.dbname = dbname
+
+        self.slab = slab
+        self.stor = self.slab.getSafeKeyVal(dbname)
+
+        if seed is None:
+            seed = s_common.guid()
+
+        self.maxusers = maxusers
+
+        self.usersbyiden = {}
```

Comment:
While we're bothering to convert these, i think it might make sense to change these to being on-demand cached lookups rather than full loads on init

Reply:
updated these to on-demand caches